### ISSUE TITLE
KeyNotFoundException thrown in Broker auth flow

### ISSUE THREAD

**Issue by martinstenhoff** [AzureAD/azure-activedirectory-library-for-dotnet#655]
We are trying to use the Microsoft Authenticator app to handle the auth flow for our xamarin ios native app (not xforms), but when our app comes back into foreground after authentication in authenticator app there's a KeyNotFoundException thrown inside the adal package (3.13.9)

We have the following setup in AAD:
```
App Service WebApp: https://ourawesomeapi.azurewebsites.net
Native App: (Has permission to our API app)
RedirectUris: ourawesomeapp://com.ourcompany.ourawesomeapp
              msauth://code/ourawesomeapp%3A%2F%2Fcom.ourcompany.ourawesomeapp
```

AppDelegate.cs
```
public override bool OpenUrl(UIApplication application, NSUrl url, string sourceApplication, NSObject annotation)
{
	if (AuthenticationContinuationHelper.IsBrokerResponse(sourceApplication))
	{
		AuthenticationContinuationHelper.SetBrokerContinuationEventArgs(url);    
	}

	return true;
}
```

AcquireTokenAsync
```
var authContext = new AuthenticationContext("https://login.microsoftonline.com/ourcompany.onmicrosoft.com");
try
{
	var result = await authContext.AcquireTokenAsync(
		"https://ourawesomeapi.azurewebsites.net", 
		<Application Id of Native AAD app>,
		"ourawesomeapp://com.ourcompany.ourawesomeapp", 
		new PlatformParameters(viewController, useBroker: true));
} catch (Exception e)
{
	// Always a KeyNotFoundException for some Dictionary used in the Broker Auth Flow inside ADAL.
	throw e;
}
```	

We call AcquireTokenAsync, which launches Authenticator app where we select the required account and log in. After that our app goes into foreground and OpenUrl method is called and we call SetBrokerContinuationEventArgs(url) and return true. After this we always get the KeyNotFoundException.

Why is there a KeyNotFoundException thrown, that seems like an internal ADAL error?

**Comment by martinstenhoff** [comment]
The issue is that refresh_token is not returned from the Microsoft Authenticator app, which causes the KeyNotFoundException to be thrown in in https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/blob/master/src/ADAL.PCL/Authority/TokenResponse.cs to throw the KeyNotFoundException.
```
Line 114:    RefreshToken = responseDictionary["refresh_token"], // <-- Throws KeyNotFoundException when using Broker
```

**Comment by kpanwar** [comment]
@martinstenhoff - Can you share your logs? I need correlation id and timestamp

**Comment by martinstenhoff** [comment]
I dumped the key value pair that is available in the dictionary which contains the correlation id, but I'm unsure of what timestamp you need.

correlation_id:B44F59DE-7A0E-43DB-ADD1-EFB719AAFEFB
authority:https%3A%2F%2Flogin.microsoftonline.com%2Fourcompany.onmicrosoft.com%2F
x-broker-app-ver:5.3.3
resource:https%3A%2F%2Fourawesomeapi.azurewebsites.net
user_id:martin%40ourcompany.com
expires_on:1493347242.158271
access_token:eyJ0eXAiOiJKV1QiLCJhbGciOi...<TRUNCATED FOR SECURITY>
client_id:<guid>
id_token:<JWT>

**Comment by kpanwar** [comment]
You can configure the logging callback to capture logs. ADAL also writes the log to NSLog. I would like you to share the ADAL logs, so that I can have someone investigate

**Comment by martinstenhoff** [comment]
The iOS developer have left for the day so we can provide the logs on Monday. 
We did confirm however that we do have successful login if we remove/comment out line 114
 ``` // RefreshToken = responseDictionary["refresh_token"],```

**Comment by martinstenhoff** [comment]
Microsoft Authenticator Incident ID: C7QKWRBE
See below for ADAL logs:
`2017-05-01 10:26:59.095 OurAppMobileiOS[1293:466868] 1/05/2017 12:26:59 AM:  - AuthenticationContext.cs: ADAL PCL.iOS with assembly version '3.13.9.1126', file version '3.13.9.1126' and informational version 'eddfa8ebda9e734efdc42ea6d41ea98bc7c998f3' is running...
2017-05-01 10:26:59.100 OurAppMobileiOS[1293:466868] Information : 1/05/2017 12:26:59 AM:  - AuthenticationContext.cs: ADAL PCL.iOS with assembly version '3.13.9.1126', file version '3.13.9.1126' and informational version 'eddfa8ebda9e734efdc42ea6d41ea98bc7c998f3' is running...
2017-05-01 10:26:59.176 OurAppMobileiOS[1293:466868] 1/05/2017 12:26:59 AM: c3f5e401-4045-4cbb-b091-d1af2ff13493 - AcquireTokenHandlerBase.cs: === Token Acquisition started:
Authority: https://login.microsoftonline.com/ourcompany.onmicrosoft.com/
Resource: https://app.ourcompanytest.p.azurewebsites.net
ClientId: 7d61adcc-7810-49a1-9911-ca0f85019077
CacheType: null
Authentication Target: User
2017-05-01 10:26:59.176 OurAppMobileiOS[1293:466868] Information : 1/05/2017 12:26:59 AM: c3f5e401-4045-4cbb-b091-d1af2ff13493 - AcquireTokenHandlerBase.cs: === Token Acquisition started:
Authority: https://login.microsoftonline.com/ourcompany.onmicrosoft.com/
Resource: https://app.ourcompanytest.p.azurewebsites.net
ClientId: 7d61adcc-7810-49a1-9911-ca0f85019077
CacheType: null
Authentication Target: User
2017-05-01 10:26:59.186 OurAppMobileiOS[1293:466868] 1/05/2017 12:26:59 AM: c3f5e401-4045-4cbb-b091-d1af2ff13493 - TokenCache.cs: Looking up cache for a token...
2017-05-01 10:26:59.186 OurAppMobileiOS[1293:466868] Verbose : 1/05/2017 12:26:59 AM: c3f5e401-4045-4cbb-b091-d1af2ff13493 - TokenCache.cs: Looking up cache for a token...
2017-05-01 10:26:59.199 OurAppMobileiOS[1293:466868] 1/05/2017 12:26:59 AM: c3f5e401-4045-4cbb-b091-d1af2ff13493 - TokenCache.cs: No matching token was found in the cache
2017-05-01 10:26:59.199 OurAppMobileiOS[1293:466868] Information : 1/05/2017 12:26:59 AM: c3f5e401-4045-4cbb-b091-d1af2ff13493 - TokenCache.cs: No matching token was found in the cache
Thread started: <Thread Pool> #3
Thread started: <Thread Pool> #4
Thread started: <Thread Pool> #5
2017-05-01 10:27:03.542 OurAppMobileiOS[1293:467060] 1/05/2017 12:27:03 AM: c3f5e401-4045-4cbb-b091-d1af2ff13493 - AcquireTokenHandlerBase.cs: System.Collections.Generic.KeyNotFoundException: The given key was not present in the dictionary.
  at System.ThrowHelper.ThrowKeyNotFoundException () [0x00000] in /Library/Frameworks/Xamarin.iOS.framework/Versions/10.8.0.175/src/mono/mcs/class/referencesource/mscorlib/system/throwhelper.cs:68 
  at System.Collections.Generic.Dictionary`2[TKey,TValue].get_Item (TKey key) [0x00021] in <54e129cad3ee4c61a6b46c3680ddc69a>:0 
  at Microsoft.IdentityModel.Clients.ActiveDirectory.TokenResponse.CreateFromBrokerResponse (System.Collections.Generic.IDictionary`2[TKey,TValue] responseDictionary) [0x0004c] in <97581c6894a642ef95d008cded8ad4ac>:0 
  at Microsoft.IdentityModel.Clients.ActiveDirectory.BrokerHelper.ResultFromBrokerResponse (System.Collections.Generic.IDictionary`2[TKey,TValue] responseDictionary) [0x00092] in <786d1e888b334ad993ac80d2bc3b6e92>:0 
  at Microsoft.IdentityModel.Clients.ActiveDirectory.BrokerHelper.ProcessBrokerResponse () [0x00093] in <786d1e888b334ad993ac80d2bc3b6e92>:0 
  at Microsoft.IdentityModel.Clients.ActiveDirectory.BrokerHelper+<AcquireTokenUsingBroker>d__8.MoveNext () [0x001a3] in <786d1e888b334ad993ac80d2bc3b6e92>:0 
--- End of stack trace from previous location where exception was thrown ---
  at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw () [0x0000c] in /Library/Frameworks/Xamarin.iOS.framework/Versions/10.8.0.175/src/mono/mcs/class/referencesource/mscorlib/system/runtime/exceptionservices/exceptionservicescommon.cs:143 
  at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Threading.Tasks.Task task) [0x00047] in /Library/Frameworks/Xamarin.iOS.framework/Versions/10.8.0.175/src/mono/mcs/class/referencesource/mscorlib/system/runtime/compilerservices/TaskAwaiter.cs:187 
  at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Threading.Tasks.Task task) [0x0002e] in /Library/Frameworks/Xamarin.iOS.framework/Versions/10.8.0.175/src/mono/mcs/class/referencesource/mscorlib/system/runtime/compilerservices/TaskAwaiter.cs:156 
  at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd (System.Threading.Tasks.Task task) [0x0000b] in /Library/Frameworks/Xamarin.iOS.framework/Versions/10.8.0.175/src/mono/mcs/class/referencesource/mscorlib/system/runtime/compilerservices/TaskAwaiter.cs:128 
  at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1+ConfiguredTaskAwaiter[TResult].GetResult () [0x00000] in <54e129cad3ee4c61a6b46c3680ddc69a>:0 
  at Microsoft.IdentityModel.Clients.ActiveDirectory.AcquireTokenHandlerBase+<RunAsync>d__55.MoveNext () [0x00403] in <97581c6894a642ef95d008cded8ad4ac>:0 
2017-05-01 10:27:03.542 OurAppMobileiOS[1293:467060] Error : 1/05/2017 12:27:03 AM: c3f5e401-4045-4cbb-b091-d1af2ff13493 - AcquireTokenHandlerBase.cs: System.Collections.Generic.KeyNotFoundException: The given key was not present in the dictionary.
  at System.ThrowHelper.ThrowKeyNotFoundException () [0x00000] in /Library/Frameworks/Xamarin.iOS.framework/Versions/10.8.0.175/src/mono/mcs/class/referencesource/mscorlib/system/throwhelper.cs:68 
  at System.Collections.Generic.Dictionary`2[TKey,TValue].get_Item (TKey key) [0x00021] in <54e129cad3ee4c61a6b46c3680ddc69a>:0 
  at Microsoft.IdentityModel.Clients.ActiveDirectory.TokenResponse.CreateFromBrokerResponse (System.Collections.Generic.IDictionary`2[TKey,TValue] responseDictionary) [0x0004c] in <97581c6894a642ef95d008cded8ad4ac>:0 
  at Microsoft.IdentityModel.Clients.ActiveDirectory.BrokerHelper.ResultFromBrokerResponse (System.Collections.Generic.IDictionary`2[TKey,TValue] responseDictionary) [0x00092] in <786d1e888b334ad993ac80d2bc3b6e92>:0 
  at Microsoft.IdentityModel.Clients.ActiveDirectory.BrokerHelper.ProcessBrokerResponse () [0x00093] in <786d1e888b334ad993ac80d2bc3b6e92>:0 
  at Microsoft.IdentityModel.Clients.ActiveDirectory.BrokerHelper+<AcquireTokenUsingBroker>d__8.MoveNext () [0x001a3] in <786d1e888b334ad993ac80d2bc3b6e92>:0 
--- End of stack trace from previous location where exception was thrown ---
  at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw () [0x0000c] in /Library/Frameworks/Xamarin.iOS.framework/Versions/10.8.0.175/src/mono/mcs/class/referencesource/mscorlib/system/runtime/exceptionservices/exceptionservicescommon.cs:143 
  at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Threading.Tasks.Task task) [0x00047] in /Library/Frameworks/Xamarin.iOS.framework/Versions/10.8.0.175/src/mono/mcs/class/referencesource/mscorlib/system/runtime/compilerservices/TaskAwaiter.cs:187 
  at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Threading.Tasks.Task task) [0x0002e] in /Library/Frameworks/Xamarin.iOS.framework/Versions/10.8.0.175/src/mono/mcs/class/referencesource/mscorlib/system/runtime/compilerservices/TaskAwaiter.cs:156 
  at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd (System.Threading.Tasks.Task task) [0x0000b] in /Library/Frameworks/Xamarin.iOS.framework/Versions/10.8.0.175/src/mono/mcs/class/referencesource/mscorlib/system/runtime/compilerservices/TaskAwaiter.cs:128 
  at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1+ConfiguredTaskAwaiter[TResult].GetResult () [0x00000] in <54e129cad3ee4c61a6b46c3680ddc69a>:0 
  at Microsoft.IdentityModel.Clients.ActiveDirectory.AcquireTokenHandlerBase+<RunAsync>d__55.MoveNext () [0x00403] in <97581c6894a642ef95d008cded8ad4ac>:0 
2017-05-01 10:27:23.602 OurAppMobileiOS[1293:467318] XPC connection interrupted
`

**Comment by kpanwar** [comment]
Thanks for the information. We will investigate.

**Comment by jasoncoolmax** [comment]
Hi @martinstenhoff, I have set up a similar broker call as shown in the logs, but I can get back the both access token and refresh token.
I need some more information from you :)
1. did you pass in `UniqueId` as user identifier type in the acquireToken call? If so, could you pass in `OptionalDisplayableId` or `RequiredDisplayableId`? As I saw some user mismatch errors in your logs.
2. did you use any external tool to modify the keychain? like remove the refresh token purposely? I saw the keychain in a strange state. If not, could you uninstall the reinstall Authenticator app and try again?

**Comment by martinstenhoff** [comment]
Which version of Microsoft Authenticator are you using? We're seeing this with 5.3.3
If you look at the first post you can see that we are not specifying any UserIdentifier in the AcquireTokenAsync call and we're using the following method in ADAL.PCL/AuthenticationContext.cs:
```
public async Task<AuthenticationResult> AcquireTokenAsync(string resource, string clientId, Uri redirectUri, IPlatformParameters parameters)
{
	return await this.AcquireTokenCommonAsync(resource, clientId, redirectUri, parameters, UserIdentifier.AnyUser).ConfigureAwait(false);
}
```

We are not knowingly modifying the keychain. The phone is MDM enrolled with AirWatch (AirWatch Agent 5.4.2), which pushes out a client certificate to the phone which is used to authenticate with Azure AD (well authentication is actually handled by SecureAuth).

1. Adding `new UserIdentifier("<useremail>", UserIdentifierType.OptionalDisplayableId)` into AcquireTokenAsync does not make a difference.
2. The iOS dev has uninstalled and installed Microsoft Authenticator multiple times.

**Comment by jasoncoolmax** [comment]
Hi @martinstenhoff, thanks for providing the sample project for reproducing the issue.

A pull request to fix Authenticator has already been made. But I don't have a timeline for when the fix will be deployed yet (after the fix, Authenticator still cannot guarantee to return refresh token in all scenarios).

And a fix is also needed for ADAL Xamarin not to crash even refresh token is not returned. You can track this issue for the progress :)

**Comment by jj09** [comment]
Is this fix in Authenticator 5.3.5 for iOS?

I still run to ```The given key was not present in the dictionary```

You can repro it with <a href="http://aka.ms/azureapp">AzureApp</a>:
1. Sign in with live account
2. Sign out
3. Try to sign in with the same live account and you get ```The given key was not present in the dictionary``` (which is being thrown by ```AuthenticationContext.AcquireTokenAsync``` method).

Workaround: go to Authenticator, remove live account you signed in with, and then you can sign in again...but you need to authenticate again.

**Comment by jennyf19** [comment]
@martinstenhoff @jj09 If you're still working with this library, can you try with the latest version [ADAL 4.4.0](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/releases/tag/ADAL-v4.4.0) and let us know if you still run into the same issue? thanks.
